### ISSUE TITLE
[FR] Separate WebSocket Service from AppFlowy Cloud Server for Improved Stability and Stateless Operation

### ISSUE THREAD

**Issue by appflowy** [AppFlowy-IO/AppFlowy-Cloud#411]
## Background
We are currently using actix-web to implement our WebSocket service within the AppFlowy cloud server infrastructure. This integration means that whenever the AppFlowy cloud server is restarted, the WebSocket service is also restarted. This restart leads to the loss of all existing WebSocket connections. To mitigate this, we have implemented auto-reconnect logic on the client side, which allows clients to automatically reconnect to the WebSocket server.

However, for long-term stability and to move towards a stateless architecture for the AppFlowy cloud server, we propose deploying the WebSocket service as a separate, standalone service. This separation would have several benefits:

1. **Improved Stability**: By decoupling the WebSocket service from the main AppFlowy cloud server, WebSocket connections would not be dropped due to the restarts of the cloud server, leading to a more stable connection experience for users.

2. **Stateless Cloud Server**: Separating the WebSocket service would be a step towards making the AppFlowy cloud server stateless, enhancing scalability and simplifying maintenance and updates.

3. **Focused Scaling and Management**: With the WebSocket service operating independently, it could be scaled based on connection load independently of the AppFlowy cloud server, allowing for more efficient resource use and focused troubleshooting and performance tuning.

## What to do
1. moving our WebSocket code from `src/biz/actix_ws` to `services/realtime`.
2. deploying the WebSocket service using Docker, making it easier to manage, scale, and isolate from other parts of our system.
3. support docker-compose-dev.yml and docker-compose.yml.

**Comment by khorshuheng** [comment]
Based on my current understanding, the websocket service API and the cloud server API share most of the code under `biz` crate. As such, if we are to move the websocket code to `services/realtime`, it's likely that `biz/collab` , `biz/snapshot` , `biz/user`, `biz/workspace` will all have to become independent libraries (or become part of the existing local libraries).

The other alternative which i can think of, is to keep the websocket code within the current directory, but build two different binaries, one for the websocket service and the other for the server API.

We could also add a command line argument / env config, which will start either a websocket service or a cloud server, depending on the argument / config value.

Which of the above options are preferable?

**Comment by appflowy** [comment]
I'm reviewing the `biz/actix_ws` module, which mainly depends on the `collab_rt` crate. Considering this dependency, it seems possible to move it to `services/realtime`. This transition would require adjustments in `api/workspace.rs` and `application.rs` to adapt to the new structure.

Once `services/realtime` is deployed as a separate service, it will focus solely on message passing. Other servers, like the original AppFlowy cloud, will communicate with the realtime service through a websocket connection or similar methods.

<img width="1293" alt="image" src="https://github.com/AppFlowy-IO/AppFlowy-Cloud/assets/86001920/d21c2550-1829-4c7b-85a6-660b5abb6588">

> The other alternative which i can think of, is to keep the websocket code within the current directory, but build two different binaries, one for the websocket service and the other for the server API
> We could also add a command line argument / env config, which will start either a websocket service or a cloud server, depending on the argument / config value.

Yes, that's correct. It appears to work as intended. To clarify, are you suggesting that we deploy different services by utilizing a configuration file to manage their deployment settings?

**Comment by khorshuheng** [comment]
> Once services/realtime is deployed as a separate service, it will focus solely on message passing.

Noted. Does that mean that we should refactor RealtimeServerActor / CollabRealtimeServer, such that they are no longer responsible for collab storage and realtime collab access control, as they currently did in https://github.com/AppFlowy-IO/AppFlowy-Cloud/blob/main/src/application.rs#L109 ? I will give it some thought to see how that can be done.

> To clarify, are you suggesting that we deploy different services by utilizing a configuration file to manage their deployment settings?

I meant something along the line of:
![Screenshot_2024-04-01_23-24-25](https://github.com/AppFlowy-IO/AppFlowy-Cloud/assets/32997938/e28c4a54-2c25-413a-be17-65664fe8501a)

Or:

`appflowy_cloud run websocket, appflowy_cloud run api`

However, if the realtime server and actor are refactored to only focusing on message passing, then it makes sense to move them completely to `services/realtime` since they won't be heavily relying on the biz crate now.

**Comment by appflowy** [comment]
> Noted. Does that mean migrating only the realtime client actor to services/realtime, then connect to the realtime server actor on the original cloud server via tungstenite?

Yes, you're right. The plan is to move the realtime client actor to the services/realtime directory, allowing it to connect to the existing realtime server actor on the cloud server through a WebSocket connection, facilitated by the tungstenite library.

Let me provide more context to you. Maybe you can come up with a better solution.Currently. the realtime client actor within the main cloud server framework, which directly handles all realtime interactions. 

![image](https://github.com/AppFlowy-IO/AppFlowy-Cloud/assets/86001920/867a53d1-a65f-4297-bd4e-7896e43ab842)

With the proposed changes, the realtime service will function independently but maintain a connection with the main cloud server

![image](https://github.com/AppFlowy-IO/AppFlowy-Cloud/assets/86001920/02d021a9-119d-40cb-b787-bb338589766f)

Beyond this initial step, we also envision migrating the CollabServer, currently implemented within the collab_rt crate, to reside within the new services/realtime service. 

![image](https://github.com/AppFlowy-IO/AppFlowy-Cloud/assets/86001920/c8777957-796d-44c0-a47e-9a78878b196b)

**Comment by khorshuheng** [comment]
The proposed changes seem reasonable to me. I will come out with a draft PR based on this architecture, then we can perform further test on it.

**Comment by appflowy** [comment]
> The proposed changes seem reasonable to me. I will come out with a draft PR based on this architecture, then we can perform further test on it.

Understood. Feel free to share if you come up with any new ideas

**Comment by appflowy** [comment]
@khorshuheng. I'm planning to implement some updates and add test cases to the actix-ws module. I'll aim to minimize any potential conflicts with your work

**Comment by khorshuheng** [comment]
Thanks, https://github.com/AppFlowy-IO/AppFlowy-Cloud/pull/453 does simplify the implementation and makes it easier for me to test the functionality. I have updated my current work accordingly.
